### PR TITLE
modify serializeEvent to be compatible with spark 3.4

### PR DESCRIPTION
**Pull request by ImpSy**

### CODE REVIEW

**Comment by lower19** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。

File: agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala [hunk 1]
```diff
@@ -230,10 +231,32 @@ class DelightStreamingConnector(sparkConf: SparkConf) extends Logging {
     * Caused by: com.databricks.spark.util.LimitedOutputStream$LimitExceededException: Exceeded 2097152 bytes (current = 2100278)
     */
   private def serializeEvent(event: SparkListenerEvent): Option[String] = {
+    val legacyMethod =
```
Comment (JoshRosen) [inline agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: Drive-by comment: the reflective method lookup is going to be hundreds of times slower than the actual method invocation.

This code will be much faster if you lift the reflection lookups out of the `def serializeEvent` method and cache the results as private fields in this object.

For example, maybe something like:

```scala
private val oldSparkEventToJsonMethod = Try(JsonProtocol.getClass.getDeclaredMethod("sparkEventToJson", classOf[SparkListenerEvent]).toOption
private val newSparkEventToJsonStringMethod = JsonProtocol.getClass.getDeclaredMethod("sparkEventToJsonString", classOf[SparkListenerEvent])
assert(oldSparkEventToJsonMethod.isDefined || newSparkEventToJsonStringMethod.isDefined, "couldn't find spark event to JSON method")

[...]

private def serializeEvent(event: SparkListenerEvent): Option[String] = {

[...]

  if (oldSparkEventToJsonMethod.isDefined) {
    Some(compact(render(oldSparkEventToJsonMethod.invoke(JsonProtocolProxy.jsonProtocol, event).asInstanceOf[JValue])))
  } else {
    Some(newSparkEventToJsonStringMethod.get
       .invoke(JsonProtocolProxy.jsonProtocol, event)
       .asInstanceOf[String]
    )
  }

[...]
```

Comment (ImpSy) [reply agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: Hey Josh 👋 
Thanks for the review, as someone who only write Scala to update this agent it's good to have feedback from a more seasoned user of the language
I updated the PR to adapt the scala pseudocode snippet if you want to take a look 😄

**Review by jrj-d — APPROVED** [review]
Great job and thanks for the testing!

File: agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala (hunk 1 above)
Comment (jrj-d) [reply agent/src/main/scala/co/datamechanics/delight/DelightStreamingConnector.scala]: Hey Josh, thanks for the heads-up about this change in Spark 3.4 and your comment in here :)

**Review by JoshRosen — APPROVED** [review]
Scala reflection changes LGTM 👍

**Comment by lower19** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。
